metricbeat/module/statsd/server/data.go

```diff
@@ -0,0 +1,193 @@
+		return s, errInvalidPacket
+	}
+
+	if len(parts) > 2 && len(parts[2]) > 0 && parts[2][0] == '@' {
+		s.sampleRate = string(parts[2][1:])
+
+		if len(parts) > 3 {
+			parts = [][]byte{parts[0], parts[1], parts[3]}
+		} else {
+			parts = [][]byte{parts[0], parts[1]}
+		}
+	}
+
+	nameSplit := bytes.SplitN(parts[0], []byte{':'}, 2)
+	if len(nameSplit) != 2 {
+		return s, errInvalidPacket
+	}
+
+	s.name = string(nameSplit[0])
+	s.value = string(nameSplit[1])
+
+	s.metricType = strings.Trim(string(parts[1]), "\n")
+
+	if len(parts) > 2 {
+		s.tags = string(parts[2])
+	}
+	return s, nil
+
+}
+
+// parse will parse a statsd metric into its components
+func parse(b []byte) ([]statsdMetric, error) {
+	metrics := []statsdMetric{}
+	for _, rawMetric := range bytes.Split(b, []byte("\n")) {
+		if len(rawMetric) > 0 {
+			metric, err := parseSingle(rawMetric)
+			if err != nil {
+				return metrics, err
+			}
+			metrics = append(metrics, metric)
+		}
+	}
+	return metrics, nil
+}
+
+func newMetricProcessor(reservoirSize int) *metricProcessor {
+	return &metricProcessor{
+		registry:      metrics.NewRegistry(),
+		reservoirSize: reservoirSize,
+	}
+}
+
+func (p *metricProcessor) processSingle(m statsdMetric) error {
+	if len(m.value) < 1 {
+		return nil
+	}
+
+	switch m.metricType {
+	case "c":
+		c := metrics.GetOrRegisterCounter(m.name, p.registry)
+		v, err := strconv.ParseInt(m.value, 10, 64)
+		if err != nil {
+			return err
+		}
+		// inc/dec or set
+		if m.value[0] == '+' || m.value[0] == '-' {
+			c.Inc(v)
+		} else {
+			c.Clear()
+			c.Inc(v)
+		}
+	case "g":
+		c := metrics.GetOrRegisterGaugeFloat64(m.name, p.registry)
+		v, err := strconv.ParseFloat(m.value, 64)
+		if err != nil {
+			return err
+		}
+		c.Update(v)
+	case "ms":
+		c := metrics.GetOrRegisterTimer(m.name, p.registry)
+		v, err := strconv.ParseFloat(m.value, 64)
+		if err != nil {
+			return err
+		}
+		c.Update(time.Millisecond * time.Duration(v))
+	case "h": // TODO: can these be floats?
+		c := metrics.GetOrRegisterHistogram(m.name, p.registry, metrics.NewUniformSample(p.reservoirSize))
+		v, err := strconv.ParseInt(m.value, 10, 64)
+		if err != nil {
+			return err
+		}
+		c.Update(v)
+	default:
+		// ignore others
+		// no support for sets, for example
+	}
+	return nil
+}
+
+func (p *metricProcessor) Process(event server.Event) error {
+	bytesRaw, ok := event.GetEvent()[server.EventDataKey]
+	if !ok {
+		return errors.New("Unable to retrieve response bytes")
+	}
+
+	b, _ := bytesRaw.([]byte)
+	if len(b) == 0 {
+		return errors.New("Request has no data")
+	}
+
+	metrics, err := parse(b)
+	if err != nil {
+		return err
+	}
+
+	for _, m := range metrics {
+		if err := p.processSingle(m); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (p *metricProcessor) GetAll() common.MapStr {
+	fields := common.MapStr{}
+	for k, v := range p.registry.GetAll() {
+		metric := common.MapStr{}
+		for mk, mv := range v {
+			metric[mk] = mv
+		}
+		fields[k] = metric
```

Comment:
For reference, this is how mv will look: https://github.com/rcrowley/go-metrics/blob/9beb055b7962d16947a14e1cd718098a2431e20e/registry.go#L127

I think it should not cause issues, as keys don't collide. I'm wondering if mapping will break for the ones containing dots, as the resulting field name will be: `statsd.my_histogram.99.9%` (notice the final dot)

Reply:
good point. Should i just replace the value with underlines or what's the standard way?

Reply:
SGTM, that's what we normally do

Reply:
is '%' a problem? replace with 'pct' ?

Reply:
It won't be a problem for Elasticsearch, I wonder if some parts of Kibana could be affected. Adding pct may be misleading too, as someone could think this value is a percentage.

How about something like `p95`, `p99`, `p99_9`

Reply:
i went for 95p etc for now as it was simple to do. What do you think?

Reply:
That sounds ok, I see the resulting metrics are like `statsd.metric4_95p`. We normally would do `statsd.metric4.95p`, for instance, for AWS Cloudwatch metrics, where the statistic is after the dot. That said... I don't have an strong opinion here

Reply:
updated 👍 thanks for the input